### ISSUE TITLE
Index exceed matrix dimension in softmaxloss function

### ISSUE THREAD

**Issue by scgyhy** [vlfeat/matconvnet#467]
1. when we extended the example code cifar10 to be applied in our application, we encountered the following problem, can anyone help us to solve this problem?

**the index c_ exceed the matrix X dimension:**

error vl_nnsoftmaxloss (line 73)
  t = Xmax + log(sum(ex,3)) - reshape(X(c_), [sz(1:2) 1 sz(4)]) ;
error vl_simplenn (line 321)
      res(i+1).x = vl_nnsoftmaxloss(res(i).x, l.class) ;
error cnn_train>process_epoch (line 313)
    res = vl_simplenn(net, im, dzdy, res, ...

2. as suggested in function vl_nnsoftmaxloss, it is deprecated and vl_nnloss is recommended instead. I have gone through the softmaxloss code in vl_nnloss function, but I could not understand how softmaxloss is calculated, can any contributor give a bit implementation details for the following code? 

switch lower(opts.loss)
  case {'log', 'softmaxlog', 'mhinge', 'mshinge'}
    % from category labels to indexes
    numPixelsPerImage = prod(inputSize(1:2)) ;
    numPixels = numPixelsPerImage * inputSize(4) ;
    imageVolume = numPixelsPerImage * inputSize(3) ;

    n = reshape(0:numPixels-1,labelSize) ;
    offset = 1 + mod(n, numPixelsPerImage) + ...
             imageVolume * fix(n / numPixelsPerImage) ;
    ci = offset + numPixelsPerImage * max(c - 1,0) ;
end

**Comment by nightrome** [comment]
That's simply the indexing into Y, because the softmax log-loss is sum(- log(Y(ci) ).

**Comment by TianCongda** [comment]
I have the same problems. Have you fixed it? If possible, could you please help you with it?

**Comment by rizwanasif** [comment]
@nightrome the last index value of ci is 207 while last value of x is at 206 .. cant figure out the issue.. help needed please

**Comment by nightrome** [comment]
Sorry, but I am not one of the authors and I never had this problem. I suggest you create a minimum working example to recreate the error so that others can try to understand it..

**Comment by rizwanasif** [comment]
@nightrome @albanie 
do you think that this error can occur with the wrong creation of imdb? because it works fine with the 21 classes for the pascal dataset for which it was programmed for.. what i did was just to use pascal2wider provided by @albanie to create my own database and then used selective search algorithm for creating probable bounding boxes .. and changed the final layer of VGG from 21 classes to 2 classes only..

**Comment by albanie** [comment]
@rizwanasif What are the values of the `labels` variable which passed to the network if you put a break point in your `fast_rcnn_train_get_batch.m` function [here](https://github.com/vlfeat/matconvnet/blob/master/examples/fast_rcnn/fast_rcnn_train_get_batch.m#L109)?

Are you able to create a script that reproduces the problem (this would make it easier to make suggestions :) )?

**Comment by rizwanasif** [comment]
@albanie the values of labels are as following when i put breakpoint at the place u mentioned.. 

size(labels) gives the result   103     1
and the values are:
labels =

     2
     2
     2
     2
     2
     2
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     2
     2
     2
     2
     2
     2
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3
     3

Do you think here is the problem? because i have only two classes and the ground truths are labelled as 1 or 2... but here its 2 and 3... dont quite get it
i didnt create a minimum working example as yet.. but i will try to use minimum data to check and share with you...

**Comment by albanie** [comment]
I believe that is the problem.  Often in detection the first label is treated as a "background" class.  If you a training a face detector in this framework, you would typically set the background label to `1` and the "face" label to `2`.  To fix your problem, the simplest thing is probably just to subtract `1` from all the labels in your `imdb`.  Just to check that it fixes the problem, try adding `labels = labels - 1` under the breakpoint line in `fast_rcnn_train_get_batch.m` and see if that runs ok.

**Comment by rizwanasif** [comment]
@albanie i think that fixes that particular problem.. thank you so much for that but it still doesnt runs and gives an error 

Error using  - 
Array dimensions must match for binary array op.

Error in dagnn.LossSmoothL1/forward (line 26)
      delta = inputs{1} - inputs{2} ;

there are total 3 cells in "inputs"
inputs{1} has size      1     1    84   103
inputs{2} has size      1     1    12   103
inputs{3} has size      1     1    12   103

is that also because the inputs{1} is considered as background and i should change that particular line to
      delta = inputs{2} - inputs{3} ; 
 ?

Thanks in advance..

**Comment by albanie** [comment]
My guess is that all three inputs should have shape `[1 1 8 103]`.  The first input can be fixed by changing your network initialisation code.  Each time that the number `84` appears in [these lines](https://github.com/vlfeat/matconvnet/blob/master/examples/fast_rcnn/fast_rcnn_init.m#L76-L80) it should be replaced by `8` (the network will make four bounding box regression predictions for each class - `4` for the background and `4` for the face class).  

The second and third input shapes can be fixed by modifying your `imdb` file property so that `imdb.classes.name = {'face'}` (rather than say, `imdb.classes.name = {'background', 'face'}`).  

This is because in the training code used [here](https://github.com/vlfeat/matconvnet/blob/master/examples/fast_rcnn/fast_rcnn_train.m#L121-L125) to compute the number of classes assumes that `background` is not included as a label on the `imdb`, so it adds one to the number of classes stored there by default.

**Comment by rizwanasif** [comment]
@albanie I understand the first part u mentioned.. the second part about modification of imdb file, i already have 2 classes in it. Infact, you can say that i named the claesses as imdb.classes.name = {'face', 'Nonface'}

Do i need to explicitly write the first class as background? 
What if i keep the imdb same and just change the nc = numel(imdb.classes.name) + 1;
 to nc = numel(imdb.classes.name);

would it achieve the same thing?

**Comment by rizwanasif** [comment]
Doing it my way have, all three inputs have the same shape as you said .. and started to train

**Comment by albanie** [comment]
Great!  Glad to hear it is running. Training on the CPU is unfortunately often very slow.

**Comment by rizwanasif** [comment]
Yes, i will shift this process to another system that has a GPU and then update you on the results.. THANK YOU SO MUCH!!!
